packages/runtime-core/__tests__/rendererOptimizedMode.spec.ts

```diff
@@ -1,3 +1,401 @@
+      '<p foo="dynamic"></p>'
+    )
+
+    propName = 'bar'
+    renderWithBlock(() => [
+      createVNode('p', { [propName]: 'dynamic' }, '', PatchFlags.FULL_PROPS)
+    ])
+
+    expect(inner(root)).toBe('<div><p bar="dynamic"></p></div>')
+    expect(block!.dynamicChildren!.length).toBe(1)
+    expect(serialize(block!.dynamicChildren![0].el as TestElement)).toBe(
+      '<p bar="dynamic"></p>'
+    )
+  })
+
+  // the order and length of the list will not change
+  test('PatchFlags: PatchFlags.STABLE_FRAGMENT', async () => {
+    let list = ['foo', 'bar']
+    render(
+      (openBlock(),
+      (block = createBlock(
+        Fragment,
+        null,
+        list.map(item => {
+          return createVNode('p', null, item, PatchFlags.TEXT)
+        }),
+        PatchFlags.STABLE_FRAGMENT
+      ))),
+      root
+    )
+
+    expect(inner(root)).toBe('<p>foo</p><p>bar</p>')
+    expect(block.dynamicChildren!.length).toBe(2)
+    expect(serialize(block.dynamicChildren![0].el as TestElement)).toBe(
+      '<p>foo</p>'
+    )
+    expect(serialize(block.dynamicChildren![1].el as TestElement)).toBe(
+      '<p>bar</p>'
+    )
+
+    list = list.map(item => item.repeat(2))
+    render(
+      (openBlock(),
+      createBlock(
+        Fragment,
+        null,
+        list.map(item => {
+          return createVNode('p', null, item, PatchFlags.TEXT)
+        }),
+        PatchFlags.STABLE_FRAGMENT
+      )),
+      root
+    )
+
+    expect(inner(root)).toBe('<p>foofoo</p><p>barbar</p>')
+    expect(block.dynamicChildren!.length).toBe(2)
+    expect(serialize(block.dynamicChildren![0].el as TestElement)).toBe(
+      '<p>foofoo</p>'
+    )
+    expect(serialize(block.dynamicChildren![1].el as TestElement)).toBe(
+      '<p>barbar</p>'
+    )
+  })
+
+  // Fragment as a block, since the `UNKEYED_FRAGMENT` flag will cause bail out of the optimization mode,
+  // so we disable tracking here.
+  test('PatchFlags: PatchFlags.UNKEYED_FRAGMENT', async () => {
+    const list = [{ tag: 'p', text: 'foo' }]
+    render(
+      (openBlock(true),
+      (block = createBlock(
+        Fragment,
+        null,
+        list.map(item => {
+          return createVNode(item.tag, null, item.text)
+        }),
+        PatchFlags.UNKEYED_FRAGMENT
+      ))),
+      root
+    )
+
+    expect(inner(root)).toBe('<p>foo</p>')
+    expect(block.dynamicChildren!.length).toBe(0)
+
+    list.unshift({ tag: 'i', text: 'bar' })
+    render(
+      (openBlock(),
+      createBlock(
+        Fragment,
+        null,
+        list.map(item => {
+          return createVNode(item.tag, null, item.text)
+        }),
+        PatchFlags.UNKEYED_FRAGMENT
+      )),
+      root
+    )
+
+    expect(inner(root)).toBe('<i>bar</i><p>foo</p>')
+    expect(block.dynamicChildren!.length).toBe(0)
+  })
+
+  // Fragment as a block, since the `KEYED_FRAGMENT` flag will cause bail out of the optimization mode,
+  // so we disable tracking here.
```

Comment:
The reason to disable tracking isn't that `KEYED_FRAGMENT` causes bail, but because this flag can only appear as a result of `v-for`, and a `v-for` Fragment always patches its children (because each child is also a block) so there's no point in tracking `dynamicChildren`.